hw9/hw9_t01.py

```diff
@@ -0,0 +1,67 @@
+"""
+Write a function that merges integer from sorted files and returns an iterator.
+
+file1.txt:
+1
+3
+5
+file2.txt:
+2
+4
+6
+list(merge_sorted_files(["file1.txt", "file2.txt"]))
+[1, 2, 3, 4, 5, 6]
+"""
+from collections import deque
+from pathlib import Path
+from typing import Iterator, List, Union
+
+
+def open_file_return_iterator(path: Union[Path, str]) -> Iterator:
+    with open(path) as file:
+        yield from (int(row) for row in file)
+
+
+def merge_2_sorted_iterators(  # noqa: C901,CCR001
+    left: Iterator, right: Iterator
+) -> Iterator:
+    result = []
+    try:
+        left_item = next(left)
+    except StopIteration:
+        result.extend(right)
+        return iter(result)
+
+    try:
+        right_item = next(right)
+    except StopIteration:
+        result.extend(left)
+        return iter(result)
+
+    while True:
+        if left_item <= right_item:
+            result.append(left_item)
+            try:
+                left_item = next(left)
+            except StopIteration:
+                result.append(right_item)
+                result.extend(right)
+                break
+        else:
+            result.append(right_item)
+            try:
+                right_item = next(right)
+            except StopIteration:
+                result.append(left_item)
+                result.extend(left)
+                break
+    return iter(result)
```

Comment:
How many items are in the result now?

Reply:
A number of items from 2 files

Reply:
Can we do better?

Reply:
We should have all items from all files at the end of our actions anyway. And pair by pair we are getting closer to this. How can we do this more efficiently?

Reply:
```python
numbers = open(...)
a = list(number)
b = list(iter(a))
```

We have `a` and `b` in memory.  `b` should be there by task definition. What are our options here?

Reply:
>  `b` should be there by task definition

Shouldn't `a` be there by the task definition?

> Write a function that merges integer from sorted files and returns an iterator.

Reply:
Is there any mention that you have to keep the whole file in the memory?  
Is there any mention that you need to store the content of files twice one in `a` and another in `b`?

Reply:
I'd like to have a couple of hints here...

Reply:
If you need to merge 2 iterators, how much data you need to keep in memory?

Reply:
I'd say 2 items. One from each at a time.

Reply:
Exactly.  What the reasonable numbers of items in the memory, if we have N iterators and requirement: each item is loaded once.